### PR TITLE
Adding SONIC ParticleNet Producer to CMSSW

### PR DESCRIPTION
**Pull request by wpmccormack**
#### PR description:

This PR introduces a producer for ParticleNet using SONIC for co-processor-enabled inferences-as-a-service.  The main versions of ParticleNet are all included (AK4, AK8 general, MassDecorrelation, and MassRegression), as are configs for both ONNX versions and PyTorch versions of the models.  For more information on the status of SONIC, please refer to this presentation to JetMET DPG: https://indico.cern.ch/event/1143469/contributions/4799423/attachments/2416540/4135308/March_28_JetMET_SONIC.pdf. Please note that the producer introduced here does not affect the main ParticleNet producer in any way (https://github.com/cms-sw/cmssw/blob/master/RecoBTag/ONNXRuntime/plugins/BoostedJetONNXJetTagsProducer.cc). This is an alternate way to run the standard ParticleNet model using inference as a service. As noted in the linked slides, this producer has been validated locally on Tier 2 resources and at scale with cloud computing, but with this PR, we hope to perform more "official" production tests.  Again, this producer does not any workflow unless explicitly called. A helper-function macro is slightly modified, as are a few config files. There will be an accompanying PR into https://github.com/cms-data/RecoBTag-Combined, which adds needed model files.

Tagging @kpedro88 @yongbinfeng @violatingcp @nhanvtran @jmduarte

#### PR validation:

This PR was tested on LPC using a Triton server running on the GPU-enabled AILab machines at FermiLab. Tests have also been performed using the ParticleNet SONIC producer at Purdue T2 and in Google Cloud. When the producer is not called, there is no performance impact, and when it is called, the Jet Tagging results are the same as for the generic ParticleNet producer.

### CODE REVIEW

**Comment by wpmccormack** [comment]
Accompanying model-file PR is here: https://github.com/cms-data/RecoBTag-Combined/pull/48

**Comment by jpata** [comment]
code checks should be addressed before we proceed further (it's worth to run `scram b code-format` every time before pushing).

**Comment by jpata** [comment]
assign btv-pog-l2 

@emilbols @cms-sw/btv-pog-l2 could you please review this from the BTV side?

**Comment by wpmccormack** [comment]
Hi @jpata , I ran `scram b code-format`, and pushed the change to one of the files.  Sorry, I think I _had_ run that command after pushing the first time, but forgot to push the change that came out of it.

**Comment by kpedro88** [comment]
test parameters:
pull_request = https://github.com/cms-data/RecoBTag-Combined/pull/48
workflow = 11824.9001,23434.9001
relvals_opt= -w standard,highstats,pileup,generator,extendedgen,production,upgrade,cleanedupgrade,ged

**Comment by kpedro88** [comment]
please test

**Comment by clacaputo** [comment]
test parameters:

- pull_request = https://github.com/cms-data/RecoBTag-Combined/pull/48
- workflow = 11824.9001,23434.9001
- relvals_opt= -w standard,highstats,pileup,generator,extendedgen,production,cleanedupgrade,ged

**Comment by clacaputo** [comment]
@cmsbuild please test

**Comment by wpmccormack** [comment]
Hi All, this recent push was made to make this PR align more closely with the deepMET PR (https://github.com/cms-sw/cmssw/pull/37963).  The particleNet model files are now not duplicated for the generic and sonic versions, and there is a symbolic link from the sonic version to the generic version

**Comment by jpata** [comment]
type btv

**Comment by clacaputo** [comment]
assign btv-pog

let's ask for explicit BTV signoff here

**Comment by johnalison** [comment]
BTV is happy to sign off.

**Review by clacaputo — CHANGES_REQUESTED** [review]
It would be good to change `cout` in `LogDebug`. I've pointed out few cases.

File: RecoBTag/ONNXRuntime/plugins/BoostedJetONNXJetTagsProducer.cc [hunk 1]
```diff
@@ -297,26 +204,27 @@ void BoostedJetONNXJetTagsProducer::make_inputs(const reco::DeepBoostedJetTagInf
...
+                                   info.lower_bound,
+                                   info.upper_bound);
+      curr_pos += insize;
       if (i == 0 && (!input_shapes_.empty())) {
-        input_shapes_[igroup][2] = val.size();
+        input_shapes_[igroup][2] = insize;
       }
 
       if (debug_) {
         std::cout << " -- var=" << varname << ", center=" << info.center << ", scale=" << info.norm_factor
```
[the remark this answers is not in the document]
Comment (wpmccormack) [reply RecoBTag/ONNXRuntime/plugins/BoostedJetONNXJetTagsProducer.cc]: Changed

File: RecoBTag/ONNXRuntime/plugins/BoostedJetONNXJetTagsProducer.cc [hunk 2]
```diff
@@ -297,26 +204,27 @@ void BoostedJetONNXJetTagsProducer::make_inputs(const reco::DeepBoostedJetTagInf
...
                   << ", replace=" << info.replace_inf_value << ", pad=" << info.pad << std::endl;
-        for (const auto &v : val) {
-          std::cout << v << ",";
+        for (unsigned i = curr_pos - insize; i < curr_pos; i++) {
+          std::cout << group_values[i] << ",";
```
[the remark this answers is not in the document]
Comment (wpmccormack) [reply RecoBTag/ONNXRuntime/plugins/BoostedJetONNXJetTagsProducer.cc]: Changed all instances here

File: RecoBTag/ONNXRuntime/plugins/ParticleNetSonicJetTagsProducer.cc [hunk 3]
```diff
@@ -0,0 +1,276 @@
...
+      src_(consumes<TagInfoCollection>(iConfig.getParameter<edm::InputTag>("src"))),
+      flav_names_(iConfig.getParameter<std::vector<std::string>>("flav_names")),
+      debug_(iConfig.getUntrackedParameter<bool>("debugMode", false)) {
+  ParticleNetConstructor(iConfig, false, input_names_, prep_info_map_, input_shapes_, input_sizes_, nullptr);
+
+  if (debug_) {
+    for (unsigned i = 0; i < input_names_.size(); ++i) {
+      const auto &group_name = input_names_.at(i);
+      if (!input_shapes_.empty()) {
+        std::cout << group_name << "\nshapes: ";
```
[the remark this answers is not in the document]
Comment (wpmccormack) [reply RecoBTag/ONNXRuntime/plugins/ParticleNetSonicJetTagsProducer.cc]: Yep, changed

File: RecoBTag/ONNXRuntime/plugins/ParticleNetSonicJetTagsProducer.cc [hunk 4]
```diff
@@ -0,0 +1,276 @@
...
+                                                  info.replace_inf_value,
+                                                  info.lower_bound,
+                                                  info.upper_bound);
+          curr_pos += insize;
+          if (i == 0 && (!input_shapes_.empty())) {
+            input_shapes_[igroup][2] = insize;
+          }
+
+          if (debug_) {
+            std::cout << " -- var=" << varname << ", center=" << info.center << ", scale=" << info.norm_factor
```
[the remark this answers is not in the document]
Comment (wpmccormack) [reply RecoBTag/ONNXRuntime/plugins/ParticleNetSonicJetTagsProducer.cc]: changed here

File: RecoBTag/ONNXRuntime/plugins/ParticleNetSonicJetTagsProducer.cc [hunk 5]
```diff
@@ -0,0 +1,276 @@
...
+        const auto &jet_ref = tag_infos->at(jet_n).jet();
+        for (std::size_t flav_n = 0; flav_n < flav_names_.size(); flav_n++) {
+          (*(output_tags[flav_n]))[jet_ref] = 0.;
+        }
+      }
+    }
+  }
+
+  if (debug_) {
+    std::cout << "=== " << iEvent.id().run() << ":" << iEvent.id().luminosityBlock() << ":" << iEvent.id().event()
```
[the remark this answers is not in the document]
Comment (wpmccormack) [reply RecoBTag/ONNXRuntime/plugins/ParticleNetSonicJetTagsProducer.cc]: Yes, changed

File: RecoBTag/ONNXRuntime/plugins/ParticleNetSonicJetTagsProducer.cc [hunk 6]
```diff
@@ -0,0 +1,276 @@
...
+      std::cout << flav_name << ", ";
+    }
+    std::cout << "\n";
+  }
+
+  // get output names from flav_names
+  for (const auto &flav_name : flav_names_) {
+    produces<JetTagCollection>(flav_name);
+  }
+  //preprocessInfoLoader(&iConfig);
```
[the remark this answers is not in the document]
Comment (wpmccormack) [reply RecoBTag/ONNXRuntime/plugins/ParticleNetSonicJetTagsProducer.cc]: Deleted, thanks!

**Comment by clacaputo** [comment]
@cmsbuild please test

**Comment by wpmccormack** [comment]
This was my bad, somehow I had forgotten to push a change in the config files to reference a new path for the model files

**Comment by clacaputo** [comment]
@cmsbuild please test

**Comment by clacaputo** [comment]
@cmsbuild please test

**Comment by clacaputo** [comment]
+reconstruction

- PR introduce an alternative way to run ParticleNet inference via SONIC (disabled by default)
- testable in the workflow = 11824.9001,23434.9001
- code reorganization does not affect existing reco
- requires https://github.com/cms-data/RecoBTag-Combined/pull/48

**Comment by clacaputo** [comment]
> BTV is happy to sign off.

Hi @johnalison , to trigger the change from `btv-pog-pending` to `btv-pog-approved`, could you please comment `+1`? Thanks

**Comment by qliphy** [comment]
ping @cms-sw/btv-pog-l2

**Comment by perrotta** [comment]
please test
(to rfresh)

**Comment by johnalison** [comment]
+1

**Comment by perrotta** [comment]
+1
- The differences reported in the test outputs are all due to a log-warning removed by #38526, which for some reason was added to this PR in the tests

**Pull request merged**